modules/ConnectionManager.py

```diff
@@ -0,0 +1,212 @@
+                else:
+                    logger.error("Problem creating login to %s", server, extra={'server': server})
+
+            handle = srv_obj.handle
+            try:
+                if handle.login(timeout=5, auto_refresh=True):
+                    self.handles[server] = handle
+                    rv = True
+            except Exception as e:
+                logger.error("Problem logging in to %s:%s", server, str(e), extra={'server': server, 'error': e})
+                rv = False
+        return rv
+
+    def update_state(self):
+        """
+        Return handles to all the servers from inventory.
+        :return: dict of handles
+                e.g. self.handles = {"server_name": "<server_handle>", ...}
+        """
+        server_list = self.get_inventory()
+        logger.debug("Create handles for Ucsm Servers: %s", server_list)
+        #import traceback
+        #traceback.print_stack()
+        # we iterate over the range to be able to modify the list in the loop body
+        rm_connections = set(self.handles.keys())
+        for sid in range(len(server_list)):
+            server = server_list[sid]
+            active = self.update_handle(server)
+            if not active:
+                continue
+            if server in rm_connections:
+                rm_connections.remove(server)
+            self.start_poll_thread(server)
+            # refresh otherwise the handle gets stale
+
+            #else:
+            #    rm_connections.remove(server)
+        # remove old connections no longer in server list
+        for s in rm_connections:
+            logger.info("remove old server connection: %s", s)
+            try:
+                del self.handles[s]
+            except: pass
+
+        return self.handles
+
+
+    def logout_handles(self):
+        """
+        Logout from all the handles
+        :return:
+        """
+        rm = []
+        for server, handle in self.handles.items():
+            logger.info("Logging out from server {}".format(server))
+            handle.logout()
+            rm.append(server)
+        for s in rm:
+            del self.handles[s]
+
+    def get_config_data(self, key=None):
+        """
+
+        :return:
+        """
+        with open(self.config['config']) as conf:
+            content = json.load(conf)
+        if key:
+            return content[key]
+        else:
+            return content
+
+    def register_collector(self, collector):
+        if collector not in self._collectors:
+            self._collectors.append(collector)
+        else:
+            logger.error("Collector %s is already registered", collector)
+
+    def get_collectors(self):
+        return self._collectors
+
+    def start_poll_thread(self, host):
+        current = self._poll_threads.get(host, None)
+        if current:
+            if not current.is_alive():
+                # thread is dead, create a new one
+                self._poll_threads.delete(host)
```

Comment:
Have you considered using a queue to check for existing threads or emptiness?

Reply:
I am working on adding queue. will create another PR.